feed_writers: optimize error path

bhalevy opened the pull request:
Eliminate one try/catch block around call to wr.close() by using coroutine::as_future.

Mark error paths as `[[unlikely]]`.

Use `coroutine::return_exception_ptr` to avoid rethrowing the final exception.

* This is an optimization.  No backport is needed

avikivity commented on mutation_writer/feed_writers.hh:
```diff
@@ -63,15 +66,16 @@ future<> feed_writer(mutation_reader&& rd_ref, Writer wr) {
+    if (ex) [[unlikely]] {
+        co_await coroutine::return_exception_ptr(std::move(ex));
     }
 }
```
In what way is this an optimization?

bhalevy replied:
IIUC, it saves the coroutine wrapper from catching the exception and converting it back to an exceptional future.
Unfortunately, the co_await'er will throw that exception, but we can't do much about it here.

bhalevy replied:
@avikivity are you OK with the patch?

avikivity replied:
Yes, but is the error path important? We never hit it.

bhalevy replied:
but what's the advantage of keeping it as it is?
just less code churn, or is there anything else, like code footprint?

avikivity replied:
Less code churn and less risk. The existing code looks simpler.

bhalevy replied:
kept `rethrow_exception` in v2 (373651c10b)

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15261/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6524/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3407/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14543/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15261/)
#### Failed Tests (7/38017):
- [test_external_dicts.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15261/testReport/junit/topology_custom/test_rpc_compression/Tests___Unit_Tests___test_external_dicts_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_external_dicts.dev.1)
- [cqlpy.cassandra_tests/validation/entities/uf_types_test.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15261/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_uf_types_test_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/uf_types_test.dev.1)
- [cqlpy.cassandra_tests/validation/entities/uf_types_test.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15261/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_uf_types_test_dev_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/uf_types_test.dev.2)
- [cqlpy.cassandra_tests/validation/entities/uf_types_test.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15261/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_uf_types_test_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/uf_types_test.dev.3)
- [cqlpy.cassandra_tests/validation/entities/uf_types_test.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15261/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_uf_types_test_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/uf_types_test.release.1)
- [cqlpy.cassandra_tests/validation/entities/uf_types_test.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15261/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_uf_types_test_release_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/uf_types_test.release.2)
- [cqlpy.cassandra_tests/validation/entities/uf_types_test.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15261/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_uf_types_test_release_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/uf_types_test.release.3)
#### Build Details:
- Duration: 9 hr 50 min
- Builder: i-064ab12d8e48f4f24 (m5ad.8xlarge)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15335/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3463/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6582/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14601/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15335/)
#### Build Details:
- Duration: 4 hr 53 min
- Builder: spider2.cloudius-systems.com

bhalevy commented:
In v2 ([373651c](https://github.com/scylladb/scylladb/commit/373651c10b2b94eb4b6a56b28c2305e3d4d661da)):
- rebased
- simplified change, keeping `rethrow_exception` on error

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15429/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3541/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6664/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14680/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15429/)
#### Build Details:
- Duration: 5 hr 1 min
- Builder: spider3.cloudius-systems.com

bhalevy commented:
@avikivity ping, please ack

avikivity commented:
Queued, but still seems gratuitous.

Is checking wr.close() even necessary? close functions shoudn't throw.

Alternatively, we should have a standard way to handle exceptions from must-run code.

```c++
    std::exception_ptr ex;
    try {
         blah blah
    } catch (...) {
         ex = std::current_exception();
    }
    // will re-raise ex or an exception from wr.close()
    co_await coroutine::with_previous_exception(std::move(ex), wr.close());
}
```

When we have [[clang::coro_await_elidable]],

```c++
    co_await coroutine::with_finally(
        [&] {
            blah blah
        },
        [&] {
             co_await wr.close();
        }
    );
```

But now it is too expensive.



/cc @tgrabiec @nyh @xemul

nyh commented:
Regarding whether close() should throw or not, I think we had a this discussion in the past, but I don't remember the conclusion ;-) 

In traditional C code, close functions **can** return error - for example, Posix fclose() and and Linux close() can return various errors in attempting to flush the output. Arguably in traditional threaded code, there wouldn't even be a point to have a close() function if it didn't return an error, because we already have a _destructor_ that can't return an error. 

The C++ standard library have a curious twist on close() returning an error: `std::ofstream::close()` is void and does not throw,... so... it does not return an error? Actually, it does: After calling fstream.close(), before destroying the object you have one last chance to check the "fail" bit to learn if the close() failed. Why did they do it this way? I guess the intention is to: suggest to the user there is no real need to call close(), a destructor will do it anyway. Yes, close() can return an error but it's an effort to check and you're not really expected to make that effort. It's perfectly fine to have the destructor implicitly close and ignore the error.

In our future-based code, when we have a future<> close() it's not optional and the user is expected to always call it. Perhaps it would have been easier if we decided that close() functions are noexcept and can't return a failed future either (and perhaps return a boolean or have a separate fail bit in a user really really cares to check). But...

There's a difficulty that any non-trivial close() function can throw exceptions unexpectedly. Surely if it tries to flush the previous writes, but even something as simple as OOM can cause it to throw. If we add noexcept, this function will crash and it will be a disaster. So to avoid this, we'll need to start catching exceptions inside close(). This will make all close implementations uglier. But maybe it's indeed better - it will make it easier to call these close() functions.

As I said above, I vaguely remember this is not the first time we've discussed this but I don't remember if we ever came up with any decision or ever documented such a decision anywhere.

avikivity commented:
> Regarding whether close() should throw or not, I think we had a this discussion in the past, but I don't remember the conclusion ;-)

Anything that needs to participate in exception recovery must not throw.


> 
> In traditional C code, close functions **can** return error - for example, Posix fclose() and and Linux close() can return various errors in attempting to flush the output. Arguably in traditional threaded code, there wouldn't even be a point to have a close() function if it didn't return an error, because we already have a _destructor_ that can't return an error.

Correct code should explicitly flush in its success path (and not rely on flushing actually happening during error recovery).


> 
> The C++ standard library have a curious twist on close() returning an error: `std::ofstream::close()` is void and does not throw,... so... it does not return an error? Actually, it does: After calling fstream.close(), before destroying the object you have one last chance to check the "fail" bit to learn if the close() failed. Why did they do it this way? I guess the intention is to: suggest to the user there is no real need to call close(), a destructor will do it anyway. Yes, close() can return an error but it's an effort to check and you're not really expected to make that effort. It's perfectly fine to have the destructor implicitly close and ignore the error.
> 
> In our future-based code, when we have a future<> close() it's not optional and the user is expected to always call it. Perhaps it would have been easier if we decided that close() functions are noexcept and can't return a failed future either (and perhaps return a boolean or have a separate fail bit in a user really really cares to check). But...

In ScyllaDB we do require that sharded::stop() and various flat_mutation_reader::close() not throw.


> 
> There's a difficulty that any non-trivial close() function can throw exceptions unexpectedly. Surely if it tries to flush the previous writes, but even something as simple as OOM can cause it to throw. If we add noexcept, this function will crash and it will be a disaster. So to avoid this, we'll need to start catching exceptions inside close(). This will make all close implementations uglier. But maybe it's indeed better - it will make it easier to call these close() functions.
> 
> As I said above, I vaguely remember this is not the first time we've discussed this but I don't remember if we ever came up with any decision or ever documented such a decision anywhere.
